File: cli/service.mjs
import { mkdir, writeFile, rm } from "node:fs/promises";
import { spawnSync } from "node:child_process";
import os from "node:os";
import path from "node:path";

/**
 * Install the always-on tap as a user login service so capture survives
 * reboots and starts on its own — the piece that makes `slink tap` truly
 * "set once." launchd on macOS, systemd --user on Linux.
 *
 * The plist/unit generators are pure (and unit-tested); install/uninstall
 * shell out to launchctl/systemctl and are only ever run by the user via
 * `slink tap --install`.
 */

export const SERVICE_LABEL = "link.session.tap";

const escapeXml = (s) =>
  String(s).replace(/[<>&'"]/g, (c) => ({ "<": "&lt;", ">": "&gt;", "&": "&amp;", "'": "&apos;", '"': "&quot;" })[c]);

/** Absolute node + CLI entry so the service never depends on the login PATH. */
export function tapProgramArgs(port = 4141) {
  return [process.execPath, process.argv[1], "tap", "--port", String(port)];
}

export function launchdPlist({ programArgs, logPath, label = SERVICE_LABEL }) {
  const args = programArgs.map((a) => `      <string>${escapeXml(a)}</string>`).join("\n");
  return `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>Label</key>
    <string>${label}</string>
    <key>ProgramArguments</key>
    <array>
${args}
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
    <key>StandardOutPath</key>
    <string>${escapeXml(logPath)}</string>
    <key>StandardErrorPath</key>
    <string>${escapeXml(logPath)}</string>
  </dict>
</plist>
`;
}

export function systemdUnit({ programArgs }) {
  const exec = programArgs.map((a) => (/\s/.test(a) ? `"${a}"` : a)).join(" ");
  return `[Unit]
Description=session.link always-on tap
After=network.target

[Service]
ExecStart=${exec}
Restart=always

[Install]
WantedBy=default.target
`;
}

function launchdPath() {
  return path.join(os.homedir(), "Library", "LaunchAgents", `${SERVICE_LABEL}.plist`);
}
function systemdPath() {
  return path.join(os.homedir(), ".config", "systemd", "user", "slink-tap.service");
}

export async function installService({ port = 4141 } = {}) {
  const programArgs = tapProgramArgs(port);
  if (process.platform === "darwin") {
    const plistPath = launchdPath();
    const logPath = path.join(os.homedir(), ".slink", "tap.log");
    await mkdir(path.dirname(plistPath), { recursive: true });
    await mkdir(path.dirname(logPath), { recursive: true });
    await writeFile(plistPath, launchdPlist({ programArgs, logPath }));
    spawnSync("launchctl", ["unload", plistPath], { stdio: "ignore" });
    const r = spawnSync("launchctl", ["load", "-w", plistPath], { encoding: "utf8" });
    return { ok: r.status === 0, path: plistPath, error: r.error?.message || r.stderr?.trim() };
  }
  if (process.platform === "linux") {
    const unitPath = systemdPath();
    await mkdir(path.dirname(unitPath), { recursive: true });
    await writeFile(unitPath, systemdUnit({ programArgs }));
    spawnSync("systemctl", ["--user", "daemon-reload"], { stdio: "ignore" });
    const r = spawnSync("systemctl", ["--user", "enable", "--now", "slink-tap.service"], { encoding: "utf8" });
    return { ok: r.status === 0, path: unitPath, error: r.error?.message || r.stderr?.trim() };
  }
  return { ok: false, error: `service install isn't supported on ${process.platform} yet` };
}

export async function uninstallService() {
  if (process.platform === "darwin") {
    const plistPath = launchdPath();
    spawnSync("launchctl", ["unload", "-w", plistPath], { stdio: "ignore" });
    await rm(plistPath, { force: true });
    return { ok: true, path: plistPath };
  }
  if (process.platform === "linux") {
    const unitPath = systemdPath();
    spawnSync("systemctl", ["--user", "disable", "--now", "slink-tap.service"], { stdio: "ignore" });
    await rm(unitPath, { force: true });
    return { ok: true, path: unitPath };
  }
  return { ok: false, error: `service uninstall isn't supported on ${process.platform} yet` };
}

File: go/internal/cli/service.go
package cli

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
)

// Login-service install for the always-on tap — launchd on macOS,
// systemd --user on Linux. The plist/unit generators are pure and
// unit-tested; install/uninstall shell out to launchctl/systemctl.

const ServiceLabel = "link.session.tap"

var xmlEscaper = strings.NewReplacer(
	"&", "&amp;", "<", "&lt;", ">", "&gt;", "'", "&apos;", `"`, "&quot;",
)

// TapProgramArgs: the Go binary is self-contained — absolute path so the
// service never depends on the login PATH.
func TapProgramArgs(port int) []string {
	exe, err := os.Executable()
	if err != nil {
		exe = "slink"
	}
	return []string{exe, "tap", "--port", fmt.Sprintf("%d", port)}
}

func LaunchdPlist(programArgs []string, logPath, label string) string {
	lines := make([]string, len(programArgs))
	for i, a := range programArgs {
		lines[i] = "      <string>" + xmlEscaper.Replace(a) + "</string>"
	}
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>Label</key>
    <string>%s</string>
    <key>ProgramArguments</key>
    <array>
%s
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
    <key>StandardOutPath</key>
    <string>%s</string>
    <key>StandardErrorPath</key>
    <string>%s</string>
  </dict>
</plist>
`, label, strings.Join(lines, "\n"), xmlEscaper.Replace(logPath), xmlEscaper.Replace(logPath))
}

var hasSpace = regexp.MustCompile(`\s`)

func SystemdUnit(programArgs []string) string {
	quoted := make([]string, len(programArgs))
	for i, a := range programArgs {
		if hasSpace.MatchString(a) {
			quoted[i] = `"` + a + `"`
		} else {
			quoted[i] = a
		}
	}
	return fmt.Sprintf(`[Unit]
Description=session.link always-on tap
After=network.target

[Service]
ExecStart=%s
Restart=always

[Install]
WantedBy=default.target
`, strings.Join(quoted, " "))
}

// ServiceResult mirrors the JS install/uninstall return.
type ServiceResult struct {
	OK    bool
	Path  string
	Error string
}

func launchdPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, "Library", "LaunchAgents", ServiceLabel+".plist")
}

func systemdPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".config", "systemd", "user", "slink-tap.service")
}

func TapLogPath() string { return filepath.Join(Home(), "tap.log") }

func InstallService(port int) ServiceResult {
	args := TapProgramArgs(port)
	switch runtime.GOOS {
	case "darwin":
		p := launchdPath()
		if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
			return ServiceResult{Error: err.Error()}
		}
		os.MkdirAll(filepath.Dir(TapLogPath()), 0o755)
		if err := os.WriteFile(p, []byte(LaunchdPlist(args, TapLogPath(), ServiceLabel)), 0o644); err != nil {
			return ServiceResult{Error: err.Error()}
		}
		exec.Command("launchctl", "unload", p).Run() // reload-safe
		if err := exec.Command("launchctl", "load", p).Run(); err != nil {
			return ServiceResult{Error: "launchctl load failed: " + err.Error(), Path: p}
		}
		return ServiceResult{OK: true, Path: p}
	case "linux":
		p := systemdPath()
		if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
			return ServiceResult{Error: err.Error()}
		}
		if err := os.WriteFile(p, []byte(SystemdUnit(args)), 0o644); err != nil {
			return ServiceResult{Error: err.Error()}
		}
		exec.Command("systemctl", "--user", "daemon-reload").Run()
		if err := exec.Command("systemctl", "--user", "enable", "--now", "slink-tap.service").Run(); err != nil {
			return ServiceResult{Error: "systemctl enable failed: " + err.Error(), Path: p}
		}
		return ServiceResult{OK: true, Path: p}
	}
	return ServiceResult{Error: "login services are supported on macOS and Linux only"}
}

func UninstallService() ServiceResult {
	switch runtime.GOOS {
	case "darwin":
		p := launchdPath()
		exec.Command("launchctl", "unload", p).Run()
		os.Remove(p)
		return ServiceResult{OK: true, Path: p}
	case "linux":
		p := systemdPath()
		exec.Command("systemctl", "--user", "disable", "--now", "slink-tap.service").Run()
		os.Remove(p)
		exec.Command("systemctl", "--user", "daemon-reload").Run()
		return ServiceResult{OK: true, Path: p}
	}
	return ServiceResult{Error: "login services are supported on macOS and Linux only"}
}
